std/assembly/index.d.ts

```diff
@@ -1314,6 +1314,10 @@ declare type valueof<T extends unknown[]> = T[0];
 declare type ReturnType<T extends (...args: any) => any> = T extends (...args: any) => infer R ? R : any;
 /** A special type evaluated to the return type of T if T is a callable function. */
 declare type returnof<T extends (...args: any) => any> = ReturnType<T>;
+/** A special type that excludes null and undefined from T. */
+declare type NonNullable<T> = T extends null | undefined ? never : T;
```

Comment:
Wait, how does this work?

Reply:
Heh, toyed around with it a little and it works, I just don't understand how

Reply:
It's a conditional type. `NonNullable<null>` becomes the `never` type which would cause it to fail to typecheck.  We don't have this issue because it's not possible to type something as `null`, [see](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1zcGVlZCZydW50aW1lPXN0dWIKLyoqIENhbGN1bGF0ZXMgdGhlIG4tdGggRmlib25hY2NpIG51bWJlci4gKi8KZXhwb3J0IGZ1bmN0aW9uIGZpYihuOiBpMzIpOiBpMzIgewogIHZhciBhID0gMCwgYiA9IDEKICBpZiAobiA+IDApIHsKICAgIHdoaWxlICgtLW4pIHsKICAgICAgbGV0IHQgPSBhICsgYgogICAgICBhID0gYgogICAgICBiID0gdAogICAgfQogICAgcmV0dXJuIGIKICB9CiAgcmV0dXJuIGEKfQpjb25zdCB4OiBudWxsID0gbnVsbDsKIyFodG1sCjx0ZXh0YXJlYSBpZD0ib3V0cHV0IiBzdHlsZT0iaGVpZ2h0OiAxMDAlOyB3aWR0aDogMTAwJSIgcmVhZG9ubHk+PC90ZXh0YXJlYT4KPHNjcmlwdD4KbG9hZGVyLmluc3RhbnRpYXRlKG1vZHVsZV93YXNtLCB7IC8qIGltcG9ydHMgKi8gfSkKICAudGhlbigoeyBleHBvcnRzIH0pID0+IHsKICAgIGNvbnN0IG91dHB1dCA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCdvdXRwdXQnKQogICAgZm9yIChsZXQgaSA9IDA7IGkgPD0gMTA7ICsraSkgewogICAgICBvdXRwdXQudmFsdWUgKz0gYGZpYigke2l9KSA9ICR7ZXhwb3J0cy5maWIoaSl9XG5gCiAgICB9CiAgfSkKPC9zY3JpcHQ+)

Reply:
My guess is that this somehow decomposes `string | null` (which extends `null` I guess?) to `string` and `null`, with the `null` turned into `never`, so that the outcome would be `string | never` with the `never` being dropped or something.